nexus_constructor/addcomponentwindow.py

```diff
@@ -0,0 +1,253 @@
+
+    def validate(self, input: str, pos: int):
+        if not input:
+            self.isValid.emit(False)
+            return QValidator.Intermediate
+        if not os.path.isfile(input):
+            self.isValid.emit(False)
+            return QValidator.Intermediate
+        for suffixes in GEOMETRY_FILE_TYPES.values():
+            for suff in suffixes:
+                if input.endswith(f".{suff}"):
+                    self.isValid.emit(True)
+                    return QValidator.Acceptable
+        self.isValid.emit(False)
+        return QValidator.Invalid
+
+    isValid = Signal(bool)
+
+
+class GeometryType(Enum):
+    NONE = 1
+    CYLINDER = 2
+    MESH = 3
+
+
+class OkValidator(QObject):
+    """
+    Validator to enable the OK button. Several criteria have to be met before this can occur depending on the geometry type.
+    """
+
+    def __init__(self, no_geometry_button, mesh_button):
+        super().__init__()
+        self.name_is_valid = False
+        self.file_is_valid = False
+        self.units_are_valid = False
+        self.no_geometry_button = no_geometry_button
+        self.mesh_button = mesh_button
+
+    def set_name_valid(self, is_valid):
+        self.name_is_valid = is_valid
+        print("Name: {}".format(self.name_is_valid))
+        self.validate_ok()
+
+    def set_file_valid(self, is_valid):
+        self.file_is_valid = is_valid
+        print("File: {}".format(self.file_is_valid))
+        self.validate_ok()
+
+    def set_units_valid(self, is_valid):
+        self.units_are_valid = is_valid
+        print("Units: {}".format(self.units_are_valid))
+        self.validate_ok()
+
+    def validate_ok(self):
+        """
+        Validates the fields in order to dictate whether the OK button should be disabled or enabled.
+        :return: None, but emits the isValid signal.
+        """
+        unacceptable = [
+            not self.name_is_valid,
+            not self.no_geometry_button.isChecked() and not self.units_are_valid,
+            self.mesh_button.isChecked() and not self.file_is_valid,
+        ]
+
+        print("Is valid {}".format(unacceptable))
+        self.isValid.emit(not any(unacceptable))
+
+    # Signal to indicate that the fields are valid or invalid. False: invalid.
+    isValid = Signal(bool)
+
+
+class AddComponentDialog(Ui_AddComponentDialog):
+    def __init__(self, nexus_wrapper: NexusWrapper):
+        super(AddComponentDialog, self).__init__()
+        self.nexus_wrapper = nexus_wrapper
+        self.geometry_model = None
+        self.component_types = make_dictionary_of_class_definitions(
+            os.path.abspath(os.path.join(os.curdir, "definitions"))
+        )
+
+    def setupUi(self, parent_dialog):
+        """ Sets up push buttons and validators for the add component window. """
+        super().setupUi(parent_dialog)
+
+        # Connect the button calls with functions
+        self.buttonBox.button(QDialogButtonBox.Ok).clicked.connect(self.on_ok)
```

Comment:
Are these camelCase names created by NeXus Designer, or can we change them to snake_case?

Reply:
They're created by Qt Designer, but I think we can change them to snake_case. I think Qt UI files are by default camelCase, not sure if there's a standard.